### PR TITLE
Add JpegCompressionDistortionGPU kernel

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It adds a new feature needed to perform JPEG-like distortion on images

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     *Created a DALI kernel to wrap the implementation details of the existing JPEG distortion CUDA kernel*
 - Affected modules and functionalities:
     *New kernel*
 - Key points relevant for the review:
     *NA*
 - Validation and testing:
     *Tests updated*
 - Documentation (including examples):
     *NA*

**JIRA TASK**: *[DALI-1932]*

### CODE REVIEW

File: dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.h [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  const uint8_t *in;  // rgb
+  uint8_t *out;  // rgb
+  ivec<2> size;
+  i64vec<2> strides;
+  mat<8, 8, uint8_t> luma_Q_table;
+  mat<8, 8, uint8_t> chroma_Q_table;
+};
+
+template <bool horz_subsample, bool vert_subsample>
+class DLL_PUBLIC JpegCompressionDistortionGPU {
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.h]: done

File: dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_test.cu [hunk 2]
```diff
@@ -50,11 +52,11 @@ cv::Mat bgr2rgb(const cv::Mat& img) {
 template <typename GTestParams>
 class JpegDistortionTestGPU : public ::testing::Test {
   using T = typename GTestParams::T;
-  static constexpr bool vert_subsample = GTestParams::vert_subsample;
-  static constexpr bool horz_subsample = GTestParams::horz_subsample;
-  static constexpr bool use_real_images = true;
-  static constexpr bool perf_run = false;
-  static constexpr bool dump_images = false;
+  const bool vert_subsample = GTestParams::vert_subsample;
+  const bool horz_subsample = GTestParams::horz_subsample;
```
Comment (mzient) [inline dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_test.cu:56]: Now that subsampling is a run-time parameters, there's no need to make them part of the type.

**Review by mzient — APPROVED** [review]
Possibly get rid of the submsampling from test's _type_ and move the to run-time param.

File: dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_test.cu (hunk 2 above)
Comment (jantonguirao) [reply dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_test.cu]: done

File: dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.h [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+#include "dali/kernels/common/block_setup.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/kernel.h"
+
+namespace dali {
+namespace kernels {
+namespace jpeg {
+
+inline float GetQualityFactorScale(int quality) {
+  quality = clamp<int>(quality, 1, 99);
```
Comment (mzient) [inline dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.h:33]:
```suggestion
  quality = clamp<int>(quality, 1, 100);
```

File: dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.h [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    {14, 17, 22, 29, 51, 87, 80, 62},
+    {18, 22, 37, 56, 68, 109, 103, 77},
+    {24, 35, 55, 64, 81, 104, 113, 92},
+    {49, 64, 78, 87, 103, 121, 120, 101},
+    {72, 92, 95, 98, 112, 100, 103, 99}
+  }};
+  auto scale = GetQualityFactorScale(quality);
+  for (int i = 0; i < 8; i++) {
+    for (int j = 0; j < 8; j++) {
+      table(i, j) = ConvertSat<uint8_t>(scale * table(i, j));
```
Comment (mzient) [inline dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.h:59]: I think it'd be better to ensure that the value is at least 1 _here_ than to check it in the kernel for every pixel in the kernel.

File: dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.h [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    {47, 66, 99, 99, 99, 99, 99, 99},
+    {99, 99, 99, 99, 99, 99, 99, 99},
+    {99, 99, 99, 99, 99, 99, 99, 99},
+    {99, 99, 99, 99, 99, 99, 99, 99},
+    {99, 99, 99, 99, 99, 99, 99, 99}
+  }};
+  auto scale = GetQualityFactorScale(quality);
+  for (int i = 0; i < 8; i++) {
+    for (int j = 0; j < 8; j++) {
+      table(i, j) = ConvertSat<uint8_t>(scale * table(i, j));
```
Comment (mzient) [inline dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.h:79]: Likewise.

**Review by mzient — APPROVED** [review]

File: dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.h [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+namespace kernels {
+namespace jpeg {
+
+inline float GetQualityFactorScale(int quality) {
+  quality = clamp<int>(quality, 1, 100);
+  float q_scale = 1.0f;
+  if (quality < 50) {
+    q_scale = 50.0f / quality;
+  } else {
+    q_scale = 2.0f - (2 * quality / 100.0f);
```
Comment (szalpal) [inline dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.h:38]:
```suggestion
    q_scale = 2.0f - (2 * quality * .001f);
```

File: dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.cu [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+void JpegCompressionDistortionGPU::Run(KernelContext &ctx, const OutListGPU<uint8_t, 3> &out,
+                                       const InListGPU<uint8_t, 3> &in, span<const int> quality) {
+  const auto &in_shape = in.shape;
+  int nsamples = in_shape.num_samples();
+  if (quality.size() > 1 && quality.size() != nsamples) {
+    throw std::invalid_argument(make_string("Received ", quality.size(),
+                                            " quality values but the batch contains ", nsamples,
+                                            " samples."));
```
Comment (szalpal) [inline dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.cu:93]: How about pointing out in the error message, that the number of quality values and batch size have to be the same? It's apparent from code, but I feel it's not obvious reading the err msg only

File: dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.cu [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+  for (int i = 0; i < nsamples; i++) {
+    auto chroma_sh = chroma_shape_.tensor_shape_span(i);
+    auto sh = in_shape.tensor_shape_span(i);
+    // used to generate logical blocks (one thread per chroma pixel)
+    chroma_sh[0] = div_ceil(sh[0], 1 + vert_subsample_);
+    chroma_sh[1] = div_ceil(sh[1], 1 + horz_subsample_);
+  }
+
+  block_setup_.SetBlockDim(dim3(32, 16, 1));
+  int xblock = 64*(2-horz_subsample_);
```
Comment (szalpal) [inline dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.cu:44]: The formatting seems to be off. Might blow up on linting ;)

File: dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_test.cu [hunk 9]
```diff
@@ -47,16 +50,18 @@ cv::Mat bgr2rgb(const cv::Mat& img) {
...
+  using T = uint8_t;
+  const bool horz_subsample;
+  const bool vert_subsample;
+  const bool use_real_images = true;
+  const bool perf_run = false;
+  const bool dump_images = false;
 
  public:
+  JpegDistortionTestGPU()
+      : horz_subsample(std::get<0>(GetParam())), vert_subsample(std::get<0>(GetParam())) {}
```
Comment (szalpal) [inline dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_test.cu:63]: Shouldn't it be ?
```suggestion
  JpegDistortionTestGPU()
      : horz_subsample(std::get<0>(GetParam())), vert_subsample(std::get<1>(GetParam())) {}
```
If it shouldn't I clearly didn't understood, what does `std::tuple<bool, bool>` denote, so maybe dropping line of documentation might come in handy

**Review by szalpal — COMMENTED** [review]
Please verify the test ctor. The rest are just suggestions.

File: dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_test.cu (hunk 9 above)
Comment (jantonguirao) [reply dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_test.cu:63]: True, that's a bug

File: dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.h (hunk 6 above)
Comment (jantonguirao) [reply dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.h:38]: this is only run once, so I prefer to keep the division, as it reads better.

**Comment by jantonguirao** [comment]
!build

**Review by szalpal — APPROVED** [review]

File: dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.cu [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+void JpegCompressionDistortionGPU::Run(KernelContext &ctx, const OutListGPU<uint8_t, 3> &out,
+                                       const InListGPU<uint8_t, 3> &in, span<const int> quality) {
+  const auto &in_shape = in.shape;
+  int nsamples = in_shape.num_samples();
+  if (quality.size() > 1 && quality.size() != nsamples) {
+    throw std::invalid_argument(
+      make_string("Unexpected number of elements in ``quality`` argument. "
+                  "The argument could contain a single value (used for the whole batch), "
+                  "one value per sample, o no values (a default is used). Received ",
```
Comment (szalpal) [inline dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.cu:94]:
```suggestion
                  "one value per sample, or no values (a default is used). Received ",
```
If the build is running already, imho this change doesn't need the CI

**Pull request merged**